test/functional/api/v1/api_keys_controller_test.rb

```diff
@@ -306,6 +306,70 @@ def self.should_expect_otp_for_update
 
       should_expect_otp_for_create
     end
+
+    context "when mfa is required" do
+      setup do
+        User.any_instance.stubs(:mfa_required?).returns true
+        authorize_with("#{@user.email}:#{@user.password}")
+      end
+
+      context "by user with mfa disabled" do
+        setup do
+          post :create, params: { name: "test-key", index_rubygems: "true" }, format: "text"
+        end
+
+        should "deny access" do
+          assert_response 403
+          mfa_error = <<~ERROR.chomp
+            For protection of your account and your gems, you are required to set up multi-factor authentication \
+            at https://rubygems.org/multifactor_auth/new.
+          ERROR
+
+          assert_match mfa_error, @response.body
+        end
+      end
+
+      context "by user on `ui_only` level" do
+        setup do
+          @user.enable_mfa!(ROTP::Base32.random_base32, :ui_only)
+          post :create, params: { name: "test-key", index_rubygems: "true" }, format: "text"
+        end
+
+        should "deny access" do
+          assert_response 403
+          mfa_error = <<~ERROR.chomp
+            For protection of your account and your gems, you are required to change your MFA level to 'UI and gem signin' or 'UI and API' \
+            at https://rubygems.org/settings/edit.
```

Comment:
Like in https://github.com/rubygems/rubygems.org/pull/3153, a link to the blog post would be helpful for users.